src/shogun/base/base_types.h

```diff
@@ -95,6 +96,39 @@ namespace shogun
 	{
 	};
 
+	template <typename...>
+	struct type_list{};
+
+	template <typename T>
+	struct type_holder
```

Comment:
anybody has any better way to do this... @gf712 @lisitsyn

Reply:
A while ago I wrote a type list struct if you want to reuse that? I think also did recursive searches like the one below but with structs.. but this looks good to me

Reply:
Btw would it be useful to have the whole inheritance tree? Or are we just doing type erasure with these classes?

Reply:
I mean what if you have `CAccuracyMeasure` which is `CAccuracyMeasure -> CContingencyTableEvaluation -> CBinaryClassEvaluation -> CEvaluation -> CSGObject`?

Reply:
mmm well i would say that the base interface classes should not overlap, or? i mean they should have one common class (CSGO) but other than that we should not have a base interface that is subset of another base interface...

Reply:
and actually this base class 'extractor' was written with this in mind... of course in that case we actually a tuple should be returned of types and those should all be exposed ;)

Reply:
yea, that could be a pain. would probably have to generate the whole thing with ctags, i was testing that a bit. And then the map increases in size.
I am just thinking that I would expect the casting to work for all the types, like it works with static and dynamic cast.
But I can't really think of a use case for that tbh..

Reply:
usually what we do is get the "base"  type, like CMachine, and use the common interface. The rest is more algorithm specific and would probably be bad practice to use?

Reply:
We have overlapping base classes atm. But in the cases I know, this is actually only due to the inability to cast any, so no need for this anymore. Eg svm&machine

Reply:
that actually depends... as this is the method that allows the parameter that is being registered as CSVM to be case able to get it as CMachine... but CSVM is exposed in the first place because of putting an svm into mkl, or? coz in that case unfortunately we need to still expose CSVM... and then this method has a problem :)

Reply:
mmm and actually the same is true for pipeline :))) ok i'll look into things.... but yeah this actually atm has a random behaviour in some edge cases :S

Reply:
actually i've dropped from sg_interface CSVM and CPipeline as in this particular case i didn't see the point having them in that set....

Reply:
I think if we can cast there is no need to have them part of the interface, so all good :)